THREESCALE-765: Brute force protection for the admin login screen

jlledom opened the pull request:
**What this PR does / why we need it**:

Add a Recaptcha to the admin portal login screen.

This PR includes a new screen under **Account Settings -> Integrate -> Bot Protection** to enable/disable the Recaptcha.

**Which issue(s) this PR fixes** 

https://issues.redhat.com/browse/THREESCALE-765

**Verification steps** 

1. Go to **Account Settings -> Integrate -> Bot Protection**
2. Select **reCAPTCHA**
3. Go to the provider login screen
4. You should see the text
```
This site is protected by reCAPTCHA and the Google Privacy Policy and Terms of Service apply.
```
5. The login screen should work as usual: return an error when the credentials are invalid and log in when the credentials are valid
6. Set `RECAPTCHA_MIN_BOT_SCORE` to `0.99` and restart the server
7. You should always be detected as a bot, no matter whether the credentials are valid or not.

**Special notes for your reviewer**:

Some references I took to write the React component:

- https://stackoverflow.com/questions/34424845/adding-script-tag-to-react-jsx
- https://w11i.me/recaptcha-v3-react
- https://stackoverflow.com/questions/53832882/react-and-recaptcha-v3
- https://developers.google.com/recaptcha/docs/v3#programmatically_invoke_the_challenge
- https://developers.google.com/recaptcha/docs/faq#id-like-to-hide-the-recaptcha-badge.-what-is-allowed

Please, @josemigallas @lvillen review the UI part, it's probably not aligned with design guidelines or not well integrated with the rest of the UI code base.

jlledom commented on spec/javascripts/utilities/ReCaptchaV3.spec.tsx:
```diff
@@ -0,0 +1,29 @@
+
+describe('ReCaptchaV3', () => {
+  it('should render itself correctly', () => {
+    const wrapper = mount(<ReCaptchaV3 enabled action="test/action" siteKey="fakeKey" />)
```
This component loads a JS script from google. @josemigallas do you know whether this test would actually download the script every time it runs?

jlledom replied:
Surprisingly I did exactly that and the cuke doesn't fail: https://github.com/3scale/porta/blob/1cbf07e991bec976bb8b9a753c5cda740410be85/features/provider/login.feature

josemigallas replied:
We probably mock kit somewhere, somehow. For javascript tests you can create a mock like like the ones for Braintree and Stripe: https://github.com/3scale/porta/blob/a25da8553ddd1b1f20f85a8a8e3988c299596f8f/spec/javascripts/__mocks__/braintree-web.js

josemigallas commented on app/javascript/src/utilities/ReCaptchaV3.tsx:
```diff
@@ -0,0 +1,42 @@
+}
+
+const ReCaptchaV3 = ({ enabled = false, siteKey, action }: Props): ReactElement => {
+  if (!enabled) return (<div>{null}</div>)
```
```suggestion
  if (!enabled) {
    return <div>{null}</div>
  }
```

Also, why `<div>{null}</div>` and not  `null`?

jlledom replied:
Because the linter fails with:

```
TS2322: Type 'null' is not assignable to type 'ReactElement<any, string | JSXElementConstructor<any>>'.
```

josemigallas commented on app/javascript/src/utilities/ReCaptchaV3.tsx:
```diff
@@ -0,0 +1,42 @@
+  action: string;
+}
+
+const ReCaptchaV3 = ({ enabled = false, siteKey, action }: Props): ReactElement => {
```
```suggestion
const ReCaptchaV3: FunctionComponent<Props> = ({ enabled = false, siteKey, action }) => {
```

josemigallas commented on app/javascript/src/utilities/ReCaptchaV3.tsx:
```diff
@@ -0,0 +1,42 @@
+
+import { useScript } from 'utilities/useScript'
+
+import type { ReactElement } from 'react'
```
```suggestion
import type { FunctionComponent } from 'react'
```

josemigallas commented on app/javascript/src/utilities/ReCaptchaV3.tsx:
```diff
@@ -0,0 +1,42 @@
+  const [token, setToken] = useState('')
+
+  useScript(`https://www.google.com/recaptcha/api.js?render=${siteKey}`, () => {
+    const grecaptcha = window.grecaptcha
```
```suggestion
    const { grecaptcha } = window
```

josemigallas commented on spec/javascripts/utilities/ReCaptchaV3.spec.tsx:
```diff
@@ -0,0 +1,29 @@
+
+const ReCaptchaV3 = jest.requireActual('utilities/ReCaptchaV3').ReCaptchaV3 as FunctionComponent<Props>
+
+describe('ReCaptchaV3', () => {
```
IMHO there's no need to "describe" the element you're in. The file is already "ReCaptchaV3". I would use describe to wrap different scenarios, for instance:
```js
describe("when recaptcha is disabled", () => {
  it('should do that...', () => {})
})

describe("when recaptcha is enabled", () => {
  it('should do this...', () => {})
})
```

josemigallas commented on features/step_definitions/provider_steps.rb:
```diff
@@ -39,6 +39,27 @@ def import_simple_layout(provider)
+  visit provider_login_path
+end
+
+Given "{provider} tries to log in" do |provider|
+  username = provider.admins.first.username
+  set_current_domain provider.external_admin_domain
+  visit provider_login_path
+  fill_in 'Email or Username', with:  username
+  fill_in 'Password', with: "supersecret"
+  click_button 'Sign in'
+end
```
If I'm not missing anything, this step is the same as https://github.com/3scale/porta/blob/3898a60f1ba8d6562a79364e215ac78ce798cea7/features/step_definitions/session_steps.rb#L3-L6

josemigallas commented on features/step_definitions/provider_steps.rb:
```diff
@@ -39,6 +39,27 @@ def import_simple_layout(provider)
   Account.find_by_domain(domain).try!(&:destroy)
 end
 
+When "{provider} wants to log in" do |provider|
+  set_current_domain provider.external_admin_domain
+  visit provider_login_path
+end
```
Instead of this use:
```cucumber
When they go to the provider login page
```

josemigallas commented on features/step_definitions/provider_steps.rb:
```diff
@@ -39,6 +39,27 @@ def import_simple_layout(provider)
+  click_button 'Sign in'
+end
+
+Then "the provider login attempt fails" do
+  assert_current_path "/p/sessions"
+  assert find_field('Email or Username')
+  assert find_field('Password')
+  assert find_button('Sign in', disabled: true)
+end
```
I would simply use:
```cucumber
And the current page is the provider login page
```
And also assert the error message (if any)
```cucumber
Then they should see "Login failed"
```

josemigallas commented on app/javascript/src/utilities/ReCaptchaV3.tsx:
```diff
@@ -0,0 +1,42 @@
+  const inputName = `g-recaptcha-response-data[${action}]`
+  const [token, setToken] = useState('')
+
+  useScript(`https://www.google.com/recaptcha/api.js?render=${siteKey}`, () => {
```
Why do you need to useEffect this? Is `siteKey` ever going to change dynamically? Maybe I am mistaken but to me it looks like `siteKey` is passed once to the component, at first render. If so, there's no need to use an effect, simply render the script.

I wonder if it would make sense to do it in the template (new.html.slim) and use a defer:
```slim
<script id="recaptcha_script" src="https://www.google.com/recaptcha/api.js?render=#{siteKey}" defer />
```

`defer` will load the script but wait for the HTML to be parsed before executing the script, but before DOMContentLoaded is triggered.

jlledom replied:
I did it because...

![I have no idea what I'm doing](https://i0.kym-cdn.com/entries/icons/original/000/008/342/ihave.jpg)

I think I'm going to keep the `useScript` component but remove the call to `useEffect`. I think that's better than just rendering `<script />` because this way I can run he callback after the script has been loaded. I'm not sure it would work the same way with `defer`

mayorova commented:
Hm, strange, it doesn't work for me as described, i.e. I don't see any Google message on the login screen, and I don't see recaptcha errors with min score 0.99 (after restarting the server).

Also, I think it's confusing that the setting in `Audience > Developer Portal > Settings > Bot Protection` controls the admin portal login :thinking: 

I think we probably need another setting for the admin portal (and its own UI for this setting). The user may decide to use recaptcha for the developer portal, and not for admin portal, and vice versa... I know that's more work, but I think it would be better.

jlledom commented:
@josemigallas I addressed all your comments on the React components. But I couldn't do the same on cucumbers. Some of your suggestions broke some other cukes, and I failed to find a way to fix it. After two days I give up.

jlledom commented:
Based on this [comment](https://issues.redhat.com/browse/THREESCALE-765?focusedId=25098196&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-25098196) on the issue, I created a new screen under **Account Settings -> Integrate -> Bot Protection** to enable the login protection for the current provider.

mayorova commented on db/migrate/20240715075900_add_admin_bot_protection_to_settings.rb:
```diff
@@ -0,0 +1,10 @@
+class AddAdminBotProtectionToSettings < ActiveRecord::Migration[6.1]
+  def up
+    add_column :settings, :admin_bot_protection_level, :string
+    change_column_default :settings, :admin_bot_protection_level, "none"
```
```suggestion
    add_column :settings, :admin_bot_protection_level, :string, default: "none"
```

Not sure if there was a specific reason to separate the two operations? If not, I think they can be merged into one.

jlledom replied:
Yes, I did that at the beginning, but I received a warning from `strong_migrations` because it's and unsafe operation, only for Oracle.

```
=== Dangerous operation detected #strong_migrations ===

Adding a column with a non-null default causes the entire table to be rewritten.
Instead, add the column without a default value, then change the default.

class AddAdminBotProtectionToSettings < ActiveRecord::Migration[6.1]
  def up
    add_column :settings, :admin_bot_protection_level, :string
    change_column_default :settings, :admin_bot_protection_level, "none"
  end

  def down
    remove_column :settings, :admin_bot_protection_level
  end
end

Then backfill the existing rows in the Rails console or a separate migration with disable_ddl_transaction!.

class BackfillAddAdminBotProtectionToSettings < ActiveRecord::Migration[6.1]
  disable_ddl_transaction!

  def up
    Setting.unscoped.in_batches do |relation| 
      relation.update_all admin_bot_protection_level: "none"
      sleep(0.01)
    end
  end
end

Then add the NOT NULL constraint in separate migrations.

/home/jlledom/redhat/porta/db/migrate/20240711145208_add_admin_bot_protection_to_settings.rb:3:in `up'

```

So I update the migration to:
- Split into to operations
- Don't backfill
- Don't make the column `NOT NULL`
- Adapt the code to interpret `NULL` as `:none`

It's much simpler, less risky and works for all DB engines.

mayorova replied:
OK, thanks for the explanation @jlledom

mayorova commented:
The feature works well, although I think that the config screen could be maybe improved a bit. It is currently not clear, what it protects, and it's even more confusing given that there is another screen in the admin portal which is identical.

The location of the dev portal one (Developer Portal > Settings > Bot Protection) kind of suggests that it's bot protection for the developer portal, but for this new screen it's not clear, and I think it can be easily confused.
Maybe it's worth adding some text that says that this bot protection is for the admin portal specifically? Not sure if it's worth mentioning that it's for login page, but probably not, as we might add more pages in future.

jlledom commented:
@mayorova I added some descriptions in https://github.com/3scale/porta/pull/3841/commits/6454b024750c40542470987f2fa5745f2005629b

I took the text from this comment in the ticket:
https://issues.redhat.com/browse/THREESCALE-765?focusedId=25146833&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-25146833

josemigallas commented on app/views/provider/admin/bot_protections/edit.html.slim:
This and sites/spam_protections/edit look the same, can we use a partial?

jlledom replied:
Done in https://github.com/3scale/porta/pull/3841/commits/2cc9c9bae2759ce9e45e60588d77a2c24cc65713

mayorova commented on app/views/provider/admin/bot_protections/_form.html.slim:
```diff
@@ -0,0 +1,11 @@
+            input_html: { disabled: !Recaptcha.captcha_configured? }
+
+    = form.actions do
+      = form.commit_button 'Update Settings'
```
no new line?

jlledom replied:
Unacceptable

mayorova commented on app/views/provider/admin/bot_protections/_form.html.slim:
```diff
@@ -0,0 +1,11 @@
+= semantic_form_for settings, url: url, html: {:id => 'bot-protection-settings' } do |form|
+  = form.inputs 'Protection against bots' do
```
With this refactor we don't keep the "admin portal" and "developer portal" label in this title. Or it was considered redundant?

Actually, we probably can move this and the commit button labels to I18n, and reuse the complete view, not only the form.

jlledom replied:
> With this refactor we don't keep the "admin portal" and "developer portal" label in this title. Or it was considered redundant?

I don't remember 🤕

> Actually, we probably can move this and the commit button labels to I18n, and reuse the complete view, not only the form.

I don't like the idea of reuse the complete view, because then we should set some parameters like `url` or `field` in the controller.

mayorova replied:
OK, but we can still use I18n for `Protection against bots` and set it to different values for admin and dev portal.

jlledom replied:
Here you go: https://github.com/3scale/porta/pull/3841/commits/1f9ef723d6f2edf877fe0e194fa43327cbe3f016

mayorova commented on app/views/provider/admin/bot_protections/_form.html.slim:
```diff
@@ -0,0 +1,11 @@
+= semantic_form_for settings, url: url, html: {:id => 'bot-protection-settings' } do |form|
+  = form.inputs t('.title', portal: portal) do
+    = form.input  field,
```
```suggestion
    = form.input field,
```

mayorova commented:
@jlledom I've updated this PR with more suggestions :grimacing: 
https://github.com/jlledom/porta/pull/2/files

I just find it easier to follow and update.

But it's OK to keep as it is, the only real suggestion is to update the description.

And also looks like this branch needs some conflict resolution.

jlledom commented:
@mayorova I applied your suggestions and resolved the conflict

mayorova commented on config/locales/en.yml:
```diff
@@ -450,6 +450,12 @@ en:
           delete_link: I understand the consequences, proceed to delete '%{name}' backend
           delete_confirmation: Are you sure you want to delete the backend '%{name}'?
+      bot_protections:
+        form:
+          title: Protection against bots on %{portal} portal
```
```suggestion
        form:
          title: Protection against bots on %{portal} portal
```

I think this is not needed anymore

jlledom replied:
Right
